File: Frontend/app/api/roleplay/conversation/route.ts
import { NextRequest, NextResponse } from 'next/server';

const GEMINI_API_KEY = process.env.GEMINI_API_KEY;

export async function POST(request: NextRequest) {
  try {
    if (!GEMINI_API_KEY) {
      return NextResponse.json(
        { error: 'Gemini API key not configured' },
        { status: 500 }
      );
    }

    const { message, conversationHistory, scenarioTitle, scenarioRole, initialPrompt, tone } = await request.json();

    if (!message) {
      return NextResponse.json(
        { error: 'Message is required' },
        { status: 400 }
      );
    }

    // Define tone instructions
    const toneInstructions = {
      Friendly: 'Be warm, encouraging, and supportive in your responses. Show enthusiasm and positivity.',
      Neutral: 'Maintain a professional and balanced demeanor. Be business-like but not cold.',
      Aggressive: 'Be challenging, skeptical, and push back on ideas. Express doubts and raise tough objections.'
    };

    const toneInstruction = tone ? toneInstructions[tone as keyof typeof toneInstructions] : toneInstructions.Neutral;

    // Build the system prompt to keep AI in character
    const systemPrompt = `You are an expert role-play simulation engine.
You are roleplaying as a ${scenarioRole} in a "${scenarioTitle}" scenario.

CRITICAL RULES - YOU MUST FOLLOW THESE:
1. STAY IN CHARACTER as the ${scenarioRole} at all times
2. NEVER break character or acknowledge you are an AI
3. NEVER provide coaching, tips, or advice to the user
4. Respond naturally as the character would in this situation
5. Ask realistic questions, raise objections, express concerns
6. Keep responses conversational and concise (2-4 sentences)
7. Show realistic emotions and reactions based on what the user says
8. If the user's pitch is unclear, express confusion or ask for clarification
9. If the user handles objections well, gradually become more interested
10. Challenge the user with realistic business concerns

CHARACTER TONE: ${toneInstruction}

Your character background: ${initialPrompt}

Respond ONLY as the ${scenarioRole}. Do not provide meta-commentary or suggestions.`;

    // Build conversation context
    const messages = [
      {
        role: 'user',
        parts: [{ text: systemPrompt }]
      }
    ];

    // Add conversation history (which now includes the current message from the frontend)
    if (conversationHistory && conversationHistory.length > 0) {
      console.log('📜 Processing conversation history:', conversationHistory.length, 'messages');
      conversationHistory.forEach((msg: any) => {
        messages.push({
          role: msg.sender === 'user' ? 'user' : 'model',
          parts: [{ text: msg.text }]
        });
      });
    }

    console.log('📤 Sending to Gemini:', messages.length, 'messages total');

    // Call Gemini API
    const response = await fetch(
      `https://generativelanguage.googleapis.com/v1beta/models/gemini-3-pro-preview:generateContent?key=${GEMINI_API_KEY}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          contents: messages,
          generationConfig: {
            temperature: 0.9,
            topK: 40,
            topP: 0.95,
            maxOutputTokens: 200,
          },
        }),
      }
    );

    if (!response.ok) {
      const errorData = await response.json();
      console.error('Gemini API error:', errorData);
      throw new Error(errorData.error?.message || 'Gemini API request failed');
    }

    const data = await response.json();
    
    // Extract the response text
    const aiResponse = data.candidates?.[0]?.content?.parts?.[0]?.text;

    if (!aiResponse) {
      throw new Error('No response from Gemini API');
    }

    return NextResponse.json({ 
      response: aiResponse,
      timestamp: new Date().toISOString()
    });

  } catch (error: any) {
    console.error('Role-play conversation error:', error);
    return NextResponse.json(
      { error: error.message || 'Failed to process conversation' },
      { status: 500 }
    );
  }
}

File: Backend/roleplay/assessment/conversation/route.py
import os
import logging
import httpx
from datetime import datetime
from fastapi import APIRouter, Request, HTTPException
from fastapi.responses import JSONResponse

router = APIRouter()

GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")

GEMINI_URL = (
    "https://generativelanguage.googleapis.com/v1beta/models/"
    "gemini-2.5-flash-lite:generateContent"
)

TIMEOUT = httpx.Timeout(60.0, connect=10.0)


@router.post("/roleplay/conversation")
async def roleplay_conversation(request: Request):
    try:
        if not GEMINI_API_KEY:
            return JSONResponse(
                content={"error": "Gemini API key not configured"},
                status_code=500,
            )

        body = await request.json()

        message = body.get("message")
        history = body.get("conversationHistory") or []
        scenario_title = body.get("scenarioTitle") or "Roleplay"
        scenario_role = body.get("scenarioRole") or "Professional"
        initial_prompt = body.get("initialPrompt") or ""
        tone = body.get("tone") or "Neutral"

        if not message:
            return JSONResponse(
                content={"error": "Message is required"},
                status_code=400,
            )

        tone_map = {
            "Friendly": "Be warm, encouraging and supportive.",
            "Neutral": "Maintain professional and balanced tone.",
            "Aggressive": "Be skeptical and challenge the user.",
        }

        tone_instruction = tone_map.get(tone, tone_map["Neutral"])

        system_prompt = f"""
You are roleplaying as {scenario_role} in "{scenario_title}".

RULES:
- Stay fully in character.
- Never break character.
- Never provide coaching.
- Respond naturally and concisely (2-4 sentences).
- Raise realistic objections when appropriate.
- Show emotional realism.

TONE: {tone_instruction}

Background:
{initial_prompt}
""".strip()

        contents = [
            {"role": "user", "parts": [{"text": system_prompt}]}
        ]

        for msg in history:
            text = msg.get("text")
            if not text:
                continue
            contents.append({
                "role": "user" if msg.get("sender") == "user" else "model",
                "parts": [{"text": text}],
            })
        # Then append current message ONCE, outside the loop
        contents.append({
            "role": "user",
            "parts": [{"text": message}],
        })


        async with httpx.AsyncClient(timeout=TIMEOUT) as client:
            response = await client.post(
                f"{GEMINI_URL}?key={GEMINI_API_KEY}",
                headers={"Content-Type": "application/json"},
                json={
                    "contents": contents,
                    "generationConfig": {
                        "temperature": 0.9,
                        "topK": 40,
                        "topP": 0.95,
                        "maxOutputTokens": 200,
                    },
                },
            )

        if response.status_code != 200:
            logging.error("Gemini error: %s", response.text)
            raise HTTPException(status_code=500, detail="Gemini request failed")

        data = response.json()

        ai_response = (
            data.get("candidates", [{}])[0]
            .get("content", {})
            .get("parts", [{}])[0]
            .get("text")
        )

        if not ai_response:
            raise HTTPException(status_code=500, detail="Empty Gemini response")

        return JSONResponse(
            content={
                "response": ai_response.strip(),
                "timestamp": datetime.utcnow().isoformat(),
            }
        )

    except Exception as e:
        logging.exception("Conversation error")
        return JSONResponse(
            content={"error": str(e)},
            status_code=500,
        )